internal/worker/dbrepl/worker.go

```diff
@@ -0,0 +1,324 @@
+
+	// WithTLSDialer returns a dbApp that can be used to connect to the Dqlite
+	// cluster.
+	WithTLSDialer(ctx context.Context) (client.DialFunc, error)
+}
+
+// DBGetter describes the ability to supply a sql.DB
+// reference for a particular database.
+type DBGetter interface {
+	// GetDB returns a sql.DB reference for the dqlite-backed database that
+	// contains the data for the specified namespace.
+	// A NotFound error is returned if the worker is unaware of the requested DB.
+	GetDB(namespace string) (database.TxnRunner, error)
+}
+
+// WorkerConfig encapsulates the configuration options for the
+// dbaccessor worker.
+type WorkerConfig struct {
+	DBGetter coredatabase.DBGetter
+	Logger   logger.Logger
+	Stdout   io.Writer
+	Stderr   io.Writer
+	Stdin    io.Reader
+}
+
+// Validate ensures that the config values are valid.
+func (c *WorkerConfig) Validate() error {
+	if c.DBGetter == nil {
+		return errors.NotValidf("missing DBGetter")
+	}
+	if c.Logger == nil {
+		return errors.NotValidf("missing Logger")
+	}
+	if c.Stdout == nil {
+		return errors.NotValidf("missing Stdout")
+	}
+	if c.Stderr == nil {
+		return errors.NotValidf("missing Stderr")
+	}
+	if c.Stdin == nil {
+		return errors.NotValidf("missing Stdin")
+	}
+	return nil
+}
+
+type dbReplWorker struct {
+	cfg  WorkerConfig
+	tomb tomb.Tomb
+
+	dbGetter coredatabase.DBGetter
+}
+
+// NewWorker creates a new dbaccessor worker.
+func NewWorker(cfg WorkerConfig) (*dbReplWorker, error) {
+	var err error
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &dbReplWorker{
+		cfg:      cfg,
+		dbGetter: cfg.DBGetter,
+	}
+
+	w.tomb.Go(w.loop)
+
+	return w, nil
+}
+
+func (w *dbReplWorker) loop() (err error) {
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	history, err := os.CreateTemp("", "juju-repl")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		_ = history.Close()
+		if err := os.Remove(history.Name()); err != nil {
+			w.cfg.Logger.Errorf("failed to remove history file: %v", err)
+		}
+	}()
+
+	line, err := readline.NewEx(&readline.Config{
+		Stdin:               readline.NewCancelableStdin(w.cfg.Stdin),
+		Stdout:              w.cfg.Stdout,
+		Stderr:              w.cfg.Stderr,
+		HistoryFile:         history.Name(),
+		InterruptPrompt:     "^C",
+		HistorySearchFold:   true,
+		FuncFilterInputRune: filterInput,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer line.Close()
+
+	currentDB, err := w.dbGetter.GetDB(database.ControllerNS)
+	if err != nil {
+		return errors.Annotate(err, "failed to get db")
+	}
+	controllerDB := currentDB
+	currentNamespace := "*"
+
+	// Allow the line to be closed when the worker is dying.
+	go func() {
+		select {
+		case <-w.tomb.Dying():
+			cancel()
+		case <-ctx.Done():
+		}
+
+		line.Close()
+	}()
+
+	// Run the main REPL loop.
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		default:
+		}
```

Comment:
Maybe this check should be moved to after we received input (`line.ReadLine()`) and before we execute the query.

This `line.ReadLine` call blocks for an indefinite amount of time, so (presumably) if a tomb dies it will be waiting on this. I assume this will then cause our query to fail for some other reason (or at least, we probably don't want the query to run).

Moving this will surely mean we're much more likely to actually catch a dying tomb?

What do you think?

Reply:
That's for your PR :wink: